data-serving/scripts/convert-data/geocode_util.py

```diff
@@ -0,0 +1,200 @@
+from typing import Any, Callable
+from constants import COMMON_LOCATION_ABBREVIATIONS
+
+
+def lookup_location(geocoder: Any, location_tokens: [str]) -> Any:
+    '''
+    Attempts to match the provided location tokens against known locations using
+    the provided geocoder.
+
+    Parameters:
+        location_tokens: One or more tokens representing a location, which 
+          may represent a part or whole of a location of any resolution.
+          Examples: ['NY'], ['Paris, France'], ['Wuhan City, Hubei, China'],
+          ['China']
+
+    Raises:
+        ValueError: When the location consists of more than 3 tokens.
+    Returns:
+      None: When no location tokens are passed in or no geocoding match is
+        found.
+      [Dict[str, Any]]: Location data of the form:
+        {
+          lat: float
+          lng: float
+          geo_resolution: str
+          country_new: str
+          admin_id: int
+          location: str
+          admin3: str
+          admin2: str
+          admin1: str
+        }
+    '''
+    normalized_tokens = [COMMON_LOCATION_ABBREVIATIONS[l.lower()]
+                         if l.lower() in COMMON_LOCATION_ABBREVIATIONS else l
+                         for l in location_tokens]
+
+    if normalized_tokens == []:
+        return None
+    if len(normalized_tokens) == 1:
+        # A location of a single token, which may be a country, city, county,
+        # etc. Ex. "Paris", "China", "FL"
+        return lookup_single_part_location(geocoder, normalized_tokens[0])
+    if len(normalized_tokens) == 2:
+        # A two-part location, which may be (county, country), (city, county),
+        # or (city, country) pairs. Ex. "Paris, France", "Hubei, China",
+        # "Boston, MA"
+        return lookup_two_part_location(
+            geocoder, normalized_tokens[0],
+            normalized_tokens[1])
+    if len(normalized_tokens) == 3:
+        # A three-part location, presumed to including city, county, and
+        # country.
+        return geocoder.geocode(
+            normalized_tokens[0],  # city
+            normalized_tokens[1],  # county
+            normalized_tokens[2])  # country
+
+    raise ValueError('Too many tokens in location')
+
+
+def lookup_two_part_location(
+        geocoder: Any, higher_res_location_token: str,
+        lower_res_location_token: str) -> Any:
+    '''
+    Attempts to match the two location tokens against known locations,
+    trying two-token pairs from broadest to most specific resolution. First we
```

Comment:
Why coarsest to finest? To avoid over-specifying in the event of a collision (on the basis that we'd rather be generally wrong than specifically wrong)? Seems weird to not error here in the event of multiple hits, TBH.

Reply:
fair enough -- done (see more details below)